libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -105,7 +106,10 @@ bool StaticAnalyzer::visit(Return const& _return)
 bool StaticAnalyzer::visit(ExpressionStatement const& _statement)
 {
 	if (_statement.expression().annotation().isPure)
-		warning(_statement.location(), "Statement has no effect.");
+		m_errorReporter.warning(_statement.location(),
```

Comment:
Perhaps I'm a relic at this point, but I'm just not a very big fan of using inheritance for mix-in behavior like that.  If there's never a need to treat a reference to `StaticAnalyzer` as a reference to `ErrorReporter`, then I don't think inheritance is the right way to go.

I did actually try to write `ErrorReporter` as a true singleton, similar to how many projects write their logger classes, but there were some components that seemed to need more real access to their `ErrorLists`, so this was a bit of a compromise.

I'm not sure that making changes to existing code smaller is completely necessary.  The changes are made here and then they're done.  That said, if there's a precedent in the project already for doing this through inheritance, then I can rewrite it that way.  Someone just needs to say the word.